src/ol/renderer/canvas/TileLayer.js

```diff
@@ -711,17 +727,35 @@ class CanvasTileLayerRenderer extends CanvasLayerRenderer {
     const wantedTiles = frameState.wantedTiles[tileSourceKey];
     const tileQueue = frameState.tileQueue;
     const minZoom = tileGrid.getMinZoom();
+    const rotation = frameState.viewState.rotation;
+    const viewport = rotation
+      ? getRotatedViewport(
+          frameState.viewState.center,
+          frameState.viewState.resolution,
+          rotation,
+          frameState.size
+        )
+      : undefined;
     let tileCount = 0;
     let tile, tileRange, tileResolution, x, y, z;
     for (z = minZoom; z <= currentZ; ++z) {
       tileRange = tileGrid.getTileRangeForExtentAndZ(extent, z, tileRange);
       tileResolution = tileGrid.getResolution(z);
       for (x = tileRange.minX; x <= tileRange.maxX; ++x) {
         for (y = tileRange.minY; y <= tileRange.maxY; ++y) {
+          const tileVisible =
+            !rotation ||
+            intersectsLinearRing(
+              viewport,
+              0,
+              viewport.length,
+              2,
+              tileGrid.getTileCoordExtent([z, x, y])
+            );
```

Comment:
Why not `continue` here when the tile is not visible?

Reply:
Thanks for your sharp eye, @MoonE. I updated the commit with your suggested improvement.